apps/server/src/modules/tool/service/tool-validation.service.spec.ts

```diff
@@ -72,27 +75,39 @@ describe('ToolValidation', () => {
 			});
 		});
 
-		describe('hasDuplicateAttributes', () => {
-			it('should not find duplicate custom parameters', async () => {
-				const externalToolDO: ExternalToolDO = externalToolDOFactory.withCustomParameters(2).build();
-
-				const result: Promise<void> = service.validateCreate(externalToolDO);
-
-				await expect(result).resolves.not.toThrow();
+		describe('hasDuplicateAttributes is called', () => {
```

Comment:
you should either write your tests in a way that you assume knowledge about the insides of your fut (whitebox test), or not assume any knowledge about the insides (blackbox tests).

When you do assume knowledge (in this case, knowing that "hasDuplicateAttributes" is an internal function), it would be better to seperate the tests for it, and test that function in isolation (via a blackbox test for only that function).

The goal is to protect ourselves from having to rewrite tests when internal structure changes, or the documentation (test descriptions!) becoming outdated.

In this case, please either move the function into a utils class and test it in isolation, or change these describe clauses. my suggestion for this case would be to have a single describe for the happy case ("when tool is valid"), and one describe for each reason to fail "when name already exists", "when there are duplicate attributes", "when regex is invalid" "when regex is missing a comment".

That way you only inform the reader about behavior, not about structure.
(do keep in mind to still be as complete as possible in each section, for example to include tests in the happy path section that include regexes "should pass with a valid regex"